src/main.rs

```diff
@@ -197,7 +197,23 @@ ___  ___      _       _         _   _      _
 
     let outputs: HashMap<String, CompileOutput> = test_sources
         .into_iter()
-        .map(|(name, entry)| (name.clone(), compiler.compile(name, entry)))
+        .map(|(name, entry)| {
+            let (in_files, out_file) = Compiler::get_paths_for(name.clone(), entry);
+
+            let output = if !Path::new(&out_file).exists()
+                || compiler::should_compile(&in_files, &out_file)
+            {
+                Log::Info(format!("Compiling {}...", name.bright_blue())).println();
+
+                compiler.compile(in_files, out_file)
+            } else {
+                Log::Info(format!("{} skipped!", name.bright_blue())).println();
+
+                compiler.skip_compile(out_file)
+            };
+
+            (name, output)
+        })
```

Comment:
I see why you would want to keep `.compile()` simple, but I feel like this is exposing functions from the `compiler` module that really will only be used in the context of `Compiler`, so they kinda pollute the API. What do you think about renaming the current `.compile()` to `.force_compile()` and creating a new "smart" `.compile()` which will handle the logic above.

Honestly, I am not totally sure. This may be just me being picky. I am open for arguments against!

Reply:
Initially I went with a wrapper method, but It was not immediately clear what happens by looking at the code just from main. But probably it was my naming, maybe I could go with an `compiler.execute` method and move the logic in `#execute`.

Reply:
imo it's fine if it doesn't get clear what compiler is doing from main. I would even argue it's better that way, it's a nice abstraction.